src/repository/opamVCS.ml

```diff
@@ -100,39 +100,65 @@ module Make (VCS: VCS) = struct
     Done (OpamStd.Option.map OpamPackage.Version.of_string r)
 
   let sync_dirty repo_root repo_url =
+    pull_url repo_root None repo_url @@+ fun result ->
     match OpamUrl.local_dir repo_url with
-    | None -> pull_url repo_root None repo_url
+    | None -> Done (result)
     | Some dir ->
+      VCS.versioned_files dir @@+ fun vc_files ->
       let files =
         List.map OpamFilename.(remove_prefix dir)
           (OpamFilename.rec_files dir)
       in
       (* Remove non-listed files from destination *)
       (* fixme: doesn't clean directories *)
       let fset = OpamStd.String.Set.of_list files in
-      List.iter (fun f ->
-          let basename = OpamFilename.remove_prefix repo_root f in
-          if not (OpamFilename.(starts_with (VCS.vc_dir repo_root) f) ||
-                  OpamStd.String.Set.mem basename fset)
-          then OpamFilename.remove f)
-        (OpamFilename.rec_files repo_root);
+      let rm_list =
+        List.fold_left (fun acc f ->
+            let basename = OpamFilename.remove_prefix repo_root f in
+            let missing =
+              not (OpamFilename.(starts_with (VCS.vc_dir repo_root) f)
+                   || OpamStd.String.Set.mem basename fset)
+            in
+            if missing then f::acc
+            else acc)
+          [] (OpamFilename.rec_files repo_root)
+      in
+      List.iter OpamFilename.remove rm_list;
+      (* We do the list cleaning here becuse of rsync options: `--exclude` need
+         to be explicitely given directory descendants, e.g `--exclude
+         _build/**`
+      *)
+      let exclude_dirs =
+        (* from [OpamGit.rsync] exclude list *)
+        let exc = [ "_opam"; "_build"; ".git"; "_darcs"; ".hg" ] in
+        fun f ->
+          List.fold_left (fun acc prefix ->
+              acc || OpamStd.String.starts_with ~prefix f)
+            false exc
+      in
+      let vcset = OpamStd.String.Set.of_list vc_files in
+      let fset =
+        OpamStd.String.Set.filter (fun f ->
+            not (OpamStd.String.Set.mem f vcset)
```

Comment:
Use set difference ?
I think it would be nicer/more readable here to get sets asap, and then focus on set operations.